### PR TITLE
[JIT][Symbolic Shape Analysis] expose op shape functions

### PR DESCRIPTION
**Pull request by makslevental**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68230

fixes https://github.com/pytorch/pytorch/issues/68204

cc @silvasean

### CODE REVIEW

**Comment by silvasean** [comment]
Thank you, thank you, thank you! This makes my life a LOT easier!

Is there a good place here to describe thoughts on the roadmap forward? Some comment somewhere? Based on what we discussed, it sounds like this is a somewhat temporary solution, but it isn't totally clear from the code.

File: torch/shape/_ops.py [hunk 1]
```diff
@@ -0,0 +1,931 @@
+R"("
```
Comment (silvasean) [inline torch/shape/_ops.py:1]: Maybe add a comment here describing that you want this while file to be a valid C++11 raw string literal, for #include'ing by C++ code.

**Comment by eellison** [comment]
Cool! before expose this in a public api (`torch.shape`) we should think a little more about what we intend the long term api to be, what the roadmap is, etc (as @silvasean asks). i'm fine landing this as a private / not stable api with something like `torch._shapes` or `torch.jit._shapes` but we should hold off on making this public for now.

**Comment by silvasean** [comment]
Does this look like it's going to land? I actually have imminent work that would benefit from this, but I can also copy these functions if needed, so no rush to get it in.

**Comment by makslevental** [comment]
> Does this look like it's going to land? I actually have imminent work that would benefit from this, but I can also copy these functions if needed, so no rush to get it in.

this can land i think. the tests that fail are primarily due to hand written tests that exercise shape analysis code (that now have slightly different values). so i just need to update those tests.

one "won't fix" failure is this one https://github.com/pytorch/pytorch/runs/4194452315 about msvc lol.

we've just both (elias and i) been backed up writing this up as a paper (not just this PR but the whole DSL). I'll try to clean up thoses tests tomorrow.

@eellison

**Pull request closed without merging**